### PR TITLE
Jgm/daos 2203 2 -- object: EC object update client changes 

### PR DESCRIPTION
**Pull request by jgmoore-or**
1. Adds parity to EC updates that include full stripes.
2. Sends update to applicable subset of shard targets (the whole set for updates that include a full stripe; parity targets, addressed targets for partial update).

### CODE REVIEW

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -1546,6 +1969,30 @@ dc_obj_update(tse_task_t *task)
 		goto out_task;
 	}
 
+	struct daos_oclass_attr *oca =
+				daos_oclass_attr_find(obj->cob_md.omd_id);
```
Comment (wangdi1) [inline src/object/cli_obj.c]: please define this with other variable at the beginning of the function. thanks.

File: src/object/cli_obj.c [hunk 2]
```diff
@@ -1509,6 +1521,415 @@ shard_task_reset_target(tse_task_t *shard_task, void *arg)
...
+		while (head != NULL) {
+			current = head;
+			D_FREE(current->niod.iod_recxs);
+			D_FREE(current->nsgl.sg_iovs);
+			for (i = 0; i < current->p_segs.p_nr; i++)
+				D_FREE(current->p_segs.p_bufs[i]);
+			D_FREE(current->p_segs.p_bufs);
+			head = current->next;
+			D_FREE(current);
+		}
```
Comment (wangdi1) [inline src/object/cli_obj.c]: seems a bit duplicate with ec_free_params_cb?

File: src/object/cli_obj.c [hunk 3]
```diff
@@ -1509,6 +1521,415 @@ shard_task_reset_target(tse_task_t *shard_task, void *arg)
...
+
+	for (i = 0; i < iod->iod_nr; i++) {
+		if (iod->iod_type == DAOS_IOD_ARRAY) {
+			int start = iod->iod_recxs[i].rx_idx * iod->iod_size;
+			int length = iod->iod_recxs[i].rx_nr * iod->iod_size;
+
+			if (length < ss) {
+				continue;
+			} else if (length - (start % ss) >= ss) {
+				*tgt_set = ~0UL;
```
Comment (wangdi1) [inline src/object/cli_obj.c]: hmm, the bit in tgt_set stands for which target the I/O will touch?  then I thought you only need set bits of (e_k+e_p)? instead of the whole bits?  or I miss sth?

Comment (jgmoore-or) [reply src/object/cli_obj.c]: This is a bit convoluted.  There are comments there to explain.  The tgt_set is initialized to zero. Non-EC updates always go to all targets, so all zeros as init value means this will happen. 

For EC updates, non-zero means full stripe, and the code in dc_obj_update sets the the tgt_set back to zero.  It is still zero if not a full stripe. Then the ec_init_tgt_set has to set correct bits.  For full stripe we can treat the target selection the same as for replica objects.  Hence, I don't need to set the correct bits for full stripe, I just set it to non-zero in the ec_update call, and use this to decide whether we need to check for specific targets. Since we don't (since it's full stripe), we set the tgt_set back to zero.  This makes the tgt selection act the same for replica objects and full-stripe updates.

File: src/object/cli_obj.c (hunk 1 above)
Comment (jgmoore-or) [reply src/object/cli_obj.c]: A (bad?) habit from years of C++ programming... I will fix.

File: src/object/cli_obj.c (hunk 2 above)
Comment (jgmoore-or) [reply src/object/cli_obj.c]: The call back has to recover the head from the data parameter via a cast from void*, and set a return code based on the task result. Otherwise it's the same code.  I can create a function that is called from both places.

File: src/object/cli_obj.c (hunk 3 above)
Comment (wangdi1) [reply src/object/cli_obj.c]: ah, so except the tgt bitmap, you also use it as a "flag". Hmm, you can set bits for both non-EC update and full stripe EC updates, which might a bit cleaner? Anyway, it is not a big deal.
Hmm, probably we should separate forward targets building for ec and non-ec update,

Comment (jgmoore-or) [reply src/object/cli_obj.c]: I was mostly interested in avoiding the pass over the recxs to determine targets if I already knew it was going to all of them.  I realized after I had it coded and tested that it was confusing, so I added the comments to explain.  I agree that it should probably be re-implemented at some point.

File: src/object/cli_obj.c [hunk 4]
```diff
@@ -1546,6 +1561,29 @@ dc_obj_update(tse_task_t *task)
...
+			 * the update should go to all targets.
+			 */
+			tgt_set = 0;
+		} else {
+			/* Called for updates with no full stripes.
+			 * Builds a bit map only if forwarding targets are
+			 * a proper subset. Sets tgt_set to zero if all targets
+			 * are addressed.
+			 */
+			ec_init_tgt_set(args->iods, args->nr, oca, &tgt_set);
```
Comment (gnailzenh) [inline src/object/cli_obj.c]: is it better to call this function within ec_obj_update?

File: src/object/cli_ec.c [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+	      uint64_t *tgt_set)
+{
+	daos_obj_update_t	*args = dc_task_get_args(task);
+	struct ec_params	*head = NULL;
+	struct ec_params	*current = NULL;
+	unsigned int		 i, j;
+	int			 rc = 0;
+
+	for (i = 0; i < args->nr; i++) {
+		if (ec_has_full_stripe(&(args->iods[i]), oca, tgt_set)) {
```
Comment (gnailzenh) [inline src/object/cli_ec.c]: style: how about declaring local variables like:
```
daos_sg_list_t *sgl = &args->sgls[i];
daos_iod_t       *iod = &args->iods[i];
```
By doing this, this code block can be easier to read?

File: src/object/cli_ec.c [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+static int
+ec_array_encode(struct ec_params *params, daos_obj_id_t oid, daos_iod_t *iod,
+		d_sg_list_t *sgl, struct daos_oclass_attr *oca,
+		int recx_idx, unsigned int *sg_idx, size_t *sg_off)
+{
+	unsigned long	s_cur;
+	unsigned int	len = oca->u.ec.e_len;
+	unsigned int	k = oca->u.ec.e_k;
+	unsigned int	p = oca->u.ec.e_p;
+	daos_recx_t     this_recx = iod->iod_recxs[recx_idx];
```
Comment (gnailzenh) [inline src/object/cli_ec.c]: no big deal, but this_recx can be a pointer right?
 ```daos_recx_t *this_recx = &iod->iod_recx[recx_idx];```

File: src/object/cli_ec.c [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+/* Allocates a stripe's worth of parity cells. */
+static int
+ec_allocate_parity(struct obj_ec_parity *par, unsigned int len, unsigned int p,
+		unsigned int prior_cnt)
+{
+	unsigned char	**nbuf;
+	unsigned int	i;
+	int		rc = 0;
+
+	D_REALLOC_ARRAY(nbuf, par->p_bufs, (prior_cnt+p));
```
Comment (gnailzenh) [inline src/object/cli_ec.c]: style: spaces around "+" please
``` prior_cnt + p```

File: src/object/cli_ec.c [hunk 8]
```diff
@@ -0,0 +1,443 @@
...
+	if (nbuf == NULL)
+		return -DER_NOMEM;
+	else if (nbuf != par->p_bufs)
+		par->p_bufs = nbuf;
+
+	for (i = prior_cnt; i < prior_cnt+p; i++) {
```
Comment (gnailzenh) [inline src/object/cli_ec.c]: same here

File: src/object/obj_rpc.h [hunk 9]
```diff
@@ -127,6 +127,7 @@ struct daos_obj_shard_tgt {
 	((uint32_t)		(orw_map_ver)		CRT_VAR) \
 	((uint32_t)		(orw_nr)		CRT_VAR) \
 	((daos_key_t)		(orw_dkey)		CRT_VAR) \
+	((uint32_t)		(orw_start_shard)	CRT_VAR) \
```
Comment (gnailzenh) [inline src/object/obj_rpc.h]: Please make sure members are 64-bit aligned, you can just move orw_flags after this line
```
uint32_t orw_map_ver;
uint32_t orw_nr;
uint32_t orw_start_shard;
uint32_t orw_flags;
daos_key_t  orw_dkey;

File: src/object/cli_ec.c [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+ * src/object/cli_ec.c
+ */
+#define D_LOGFAC	DD_FAC(object)
+
+#include <daos_task.h>
+#include <daos_types.h>
+#include "obj_rpc.h"
+#include "obj_internal.h"
+
+#define D_LOGFAC        DD_FAC(object)
```
Comment (liuxuezhao) [inline src/object/cli_ec.c]: duplicated with L29

File: src/object/cli_ec.c [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+ */
+void
+ec_init_tgt_set(daos_iod_t *iods, unsigned int nr,
+		struct daos_oclass_attr *oca, unsigned long *tgt_set)
+{
+	unsigned int    len = oca->u.ec.e_len;
+	unsigned int    k = oca->u.ec.e_k;
+	unsigned int    p = oca->u.ec.e_p;
+	unsigned long	par_only = (1UL << p) - 1;
+	unsigned long	full = (1UL <<  (k+p)) - 1;
```
Comment (liuxuezhao) [inline src/object/cli_ec.c]: why some place use uint64_t and some other place use unsigned long? all place with uint64_t would be better.

And, could we assume (k+p) <= 64?  probably cannot.

File: src/object/obj_rpc.h (hunk 9 above)
Comment (liuxuezhao) [reply src/object/obj_rpc.h]: why need to add orw_start_shard?
for current full stripe update, shard_shard is always 0, no? it it is true, we probably need not worry about it now and leave it to later.

File: src/object/cli_ec.c (hunk 11 above)
Comment (jgmoore-or) [reply src/object/cli_ec.c]: It was established in another context that k <= 64, although that context may no longer be valid.  I think it's safe to assume k+p <= 64, but I am willing to discuss it.  (I've not seen a proposal for k > 16 and/or p > 3.

Comment (liuxuezhao) [reply src/object/cli_ec.c]: > It was established in another context that k <= 64, although that context may no longer be valid. I think it's safe to assume k+p <= 64, but I am willing to discuss it. (I've not seen a proposal for k > 16 and/or p > 3.

GF(2^8) allows k/p's max value as 255.

Personally i think probably unnecessary to add (k+p <= 64) restriction if only for this tgt_set bitmap usage. May refer rebuild's rgt_scan_bits/rgt_pull_bits usage.

On the other hand, I think it not bad to add restriction of (k + p) <= 64 if the real use cases don't need more, But I think it is better to explain it at API header file or some readme file, and check it within obj_ec_codec_init() and return error & err msg for larger EC object class. Without any checking/warning and with internal restriction that will lead to silent failure seems not good.

File: src/object/obj_rpc.h (hunk 9 above)
Comment (jgmoore-or) [reply src/object/obj_rpc.h]: Start shard is not always zero. This is why I added the RPC parameter.  We talked about this in a meeting, and I recall that everyone (you, WangDi, Liang, Johann) all agreed that I should add this.

Without it, the targets cannot determine where they are in the stripe.

File: src/object/cli_ec.c [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+/* Iterates over the IODs in the update, encoding all full stripes contained
+ * within each recx.
+ */
+int
+ec_obj_update(tse_task_t *task, daos_obj_id_t oid, daos_oclass_attr_t *oca,
+	      uint64_t *tgt_set)
+{
+	daos_obj_update_t	*args = dc_task_get_args(task);
+	struct ec_params	*head = NULL;
+	struct ec_params	*current = NULL;
```
[the remark this answers is not in the document]
Comment (jgmoore-or) [reply src/object/cli_ec.c]: Mohamad does the same param list processing in dac_array, and he uses "current" -- I'd like to stick with this name for consistency.

File: src/object/obj_rpc.h (hunk 9 above)
Comment (liuxuezhao) [reply src/object/obj_rpc.h]: Sorry I am still not very clear why we need orw_start_shard.
I think each target need to know the offset within the full stripe, for example for e_k == 8, then each target need to know it is the 0, or 1, 2... 7 target in the full stripe.
When server process the update request, by check object layout it can know its shard_number, and (shard_number / e_k) is the offset it needs to know. Why still need that orw_start_shard?

Comment (jgmoore-or) [reply src/object/obj_rpc.h]: This was also discussed in the meeting held on 3/7/2019 at 7:00am MST.  The entire team (you, WangDi, Liang, Johann) approved this approach and agreed that the start shard should be sent in the RPC.  There was not dissent or disagreement from anyone in that meeting.

I had looked at the object layout prior to the 3/7/19 meeting and I determined that the function calls available to determine the object layout were all designed to be run within the client, and hence this approach would require dc_ (client based) representations of the object that are not used with the server. Because of this difficulty, everyone in the meeting agreed that just sending the start shard in the RPC would be preferred approach. 

To summarize, the shard number sent in the RPC is not indexed based on the redundancy group and cannot be mapped to 0...k-1 without knowing the index of the starting shard in the redundancy group.  Although it is probably possible to determine the index of the starting shard in the redundancy group on the server by somehow looking this value up in the object layout, it was decided that since the client already has this information, that it should just be sent in the RPC.

Comment (liuxuezhao) [reply src/object/obj_rpc.h]: Thanks for explanation. I do remember we ever discussed that, but in my memory the approach was that when leader server dispatch the IO req to different target, the leader will tell each target the shard index of that target. 
So this orw_start_shard is for the client tell the leader the shart_shart, so when leader dispatch it can calculate the shard offset/index and make it known for the non-leader targets? is it what you plan to do? 
just that did not see other related code in this patch, so just to confirm ...

Comment (jgmoore-or) [reply src/object/obj_rpc.h]: The meeting was a long time ago, so it's understandable that you don't remember the discussion.

The leader does tell each target it's shard_index, but this is not sufficient (since it's not zero-relative) to know position in the stripe.  There is no EC-specific handling planned for the leader (WangDi insists on this). Instead, the start_shard is forwarded in the RPC call to the target (the leader just forwards the orw received from the client and only changes a few parameters (e.g., shard_index).  The targets will then compute shard_index - start_shard to determine their position in the stripe.
